src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -943,46 +993,63 @@
        (when include-hash?
          {:hash (sutils/munge-hash-for-storage
                  (shash/fact-identity-hash fact-data))})))
-    (let [paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
-          pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
+    (let [paths-and-values (facts/facts->paths-and-values values)
+          pathstrs (map (comp facts/factpath-to-string first) paths-and-values)
           paths->ids (realize-paths! pathstrs)
-          valuemaps (map second paths-and-valuemaps)]
+          path-values (map second paths-and-values)]
       (insert-facts! (certname-to-factset-id certname)
                      (map paths->ids pathstrs)
-                     valuemaps)))))
+                     path-values)))))
 
 (defn- update-existing-values [factset-id change-info]
   (when (seq change-info)
     (sql/db-do-prepared
      jdbc/*db* true
      (cons
       (str "update facts set"
-           "    value_type_id = ?,"
-           "    value_boolean = ?,"
-           "    value_integer = ?,"
-           "    value_float = ?,"
-           "    value_string = ?,"
-           "    value = ?,"
-           "    large_value_hash = ?"
+           (->> (for [c value-row-cols] (str " " (name c) " = ?"))
+                (str/join ","))
            "  where factset_id = ? and fact_path_id = ?")
-      (for [[_ path path-id new-vm] change-info]
-        [(:value_type_id new-vm)
-         (:value_boolean new-vm)
-         (:value_integer new-vm)
-         (:value_float new-vm)
-         (:value_string new-vm)
-         (:value new-vm)
-         (some-> (:large_value_hash new-vm) sutils/munge-hash-for-storage)
-         factset-id
-         path-id]))
+      (for [[_ path path-id new-val new-hash] change-info]
+        (conj (if new-hash
+                (value->value-row new-val new-hash)
+                (value->value-row new-val))
+              factset-id
+              path-id)))
      {:multi? true})))
 
-(defn fact-value-changed?
-  [row new-valuemap]
-  (or (not= (:value_type_id row) (:value_type_id new-valuemap))
-      (when-let [row-hash (:large_value_hash row)]
-        (not= row-hash (:large_value_hash new-valuemap)))
-      (not= (:value row) (:value new-valuemap))))
+(defn fact-value-row-changes
+  "Returns logically false if new-val does not differ from the row data,
+  logically true otherwise, and the true value will be a hash of
+  new-val if one was computed during the evaluation."
+  [new-val type-id hash v vstr vint vfloat vbool]
+  (or (not= type-id (fact-value-type-id new-val))
+      (cond
+        ;; faster direct compare
+        (or (= type-id db-fv-int)
```

Comment:
It might make this block more readable to name these predicates. Maybe this is `primitive-fact-value?`, below is `nil-fact-value?` and below that is `big-fact-value?`. That might help take some of the noise away here and improve clarity